cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp

```diff
@@ -154,66 +154,58 @@ sycl::event kernels_fp<Float>::select(sycl::queue& queue,
-
-                    std::int32_t index = -1;
-                    Float value = fp_max;
-                    for (std::int64_t i = local_id; i < cluster_count_as_int32; i += local_range) {
-                        const Float cur_val =
-                            distances_ptr[in_offset + i] + centroid_squares_ptr[i];
-                        if (cur_val < value) {
-                            index = static_cast<std::int32_t>(i);
-                            value = cur_val;
-                        }
-                    }
 
-                    sg.barrier();
-
-                    const Float final_value =
-                        sycl::reduce_over_group(sg, value, sycl::ext::oneapi::minimum<Float>());
-                    const bool present = (final_value == value);
-                    const std::int32_t pos =
-                        sycl::exclusive_scan_over_group(sg,
-                                                        present ? 1 : 0,
-                                                        sycl::ext::oneapi::plus<std::int32_t>());
-                    const bool owner = present && pos == 0;
-                    const std::int32_t final_index =
-                        -sycl::reduce_over_group(sg,
-                                                 owner ? -index : 1,
-                                                 sycl::ext::oneapi::minimum<std::int32_t>());
+            cgh.parallel_for<select_min_distance<Float>>(range, [=](sycl::nd_item<2> item) {
+                constexpr sycl::ext::oneapi::minimum<Float> minimum_val;
+                constexpr sycl::ext::oneapi::minimum<std::int64_t> minimum_idx;
 
-                    if (local_id == 0) {
-                        indices_ptr[out_offset] = final_index;
-                        selection_ptr[out_offset] = final_value;
-                    }
-                });
+                const std::int64_t row = item.get_global_id(0) + first_row;
+
+                if (last_row <= row)
+                    return;
+                auto sg = item.get_sub_group();
+                std::int64_t local_id = item.get_local_id(1);
+                auto min_val = std::numeric_limits<Float>::max();
+                auto min_idx = std::numeric_limits<std::int64_t>::max();
+
+                const auto* const row_ptr = distances_ptr + row * stride;
+                for (std::int64_t col = local_id; col < cluster_count_as_int32; ++col) {
+                    const Float cur_val = row_ptr[col] + centroid_squares_ptr[col];
+                    const bool handle = cur_val < min_val;
+                    min_val = handle ? cur_val : min_val;
+                    min_idx = handle ? col : min_idx;
+                }
+
+                sg.barrier();
+
+                const auto true_min_val = sycl::reduce_over_group(sg, min_val, minimum_val);
+                const auto handle =
+                    (min_val == true_min_val) ? min_idx : std::numeric_limits<std::int64_t>::max();
```

Comment:
This value is `std::int32_t` in all [previous implementations](https://github.com/oneapi-src/oneDAL/blob/fbe7020d8d8de9ced5a047dcb9d35f2866184a94/cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_fp_impl.hpp#L156).

Reply:
```suggestion
                    (min_val == true_min_val) ? min_idx : std::numeric_limits<std::int32_t>::max();
```